### PR TITLE
Add ability to specify custom headers in constructor

### PR DESCRIPTION
**Pull request by Mario54**
@lemieux @dannytranlx @mspensieri 

We can now pass in custom HTTP headers in the constructor by doing
```js
new Smooch(..., {
    headers: {
        myCustomHeader: '...'
    }
});
```

To pass in the service url, it becomes:
```js
new Smooch(..., {
    serviceUrl: '...'
    headers: {
        myCustomHeader: '...'
    }
});
```

### CODE REVIEW

File: src/api/base.js [hunk 1]
```diff
@@ -57,6 +59,25 @@ export class BaseApi {
             return Promise.reject(new Error('Must not use an app token for authentication.'));
         }
 
-        return Promise.resolve(this.authHeaders);
+        return Promise.resolve();
+    }
+
+
+    request(method, url, data, {allowedAuth=this.allowedAuth} = {}) {
```
Comment (lemieux) [inline src/api/base.js]: nice

File: src/api/base.js [hunk 2]
```diff
@@ -57,6 +59,25 @@ export class BaseApi {
...
+            });
+    }
+
+    /**
+     * Combines authorization headers and custom headers passed in the Smooch constructor
+     * returns {object} - The headers to be sent in HTTP requests
+     */
+    getHeaders() {
+        return {
+            ...this.authHeaders,
```
Comment (lemieux) [inline src/api/base.js]: put headers first so that auth headers are not overriden

Comment (Mario54) [inline src/api/base.js]: Aw, crap. I meant to do that

File: src/smooch.js [hunk 3]
```diff
@@ -7,7 +7,8 @@ import packageInfo from '../package.json';
 export const SERVICE_URL = 'https://api.smooch.io/v1';
 
 export class Smooch {
-    constructor(auth = {}, serviceUrl = SERVICE_URL) {
+    constructor(auth = {}, options = {}) {
+        const {serviceUrl = SERVICE_URL, headers = {}} = options;
```
Comment (lemieux) [inline src/smooch.js]: you could do that right in the param definition `{serviceUrl, headers} = {serviceUrl = SERVICE_URL, headers = {}}` but okay

Comment (lemieux) [inline src/smooch.js]: or I guess one or the other wouldn't be provided

**Comment by lemieux** [comment]
:+1:

File: src/smooch.js (hunk 3 above)
Comment (Mario54) [inline src/smooch.js]: It was a lot to do on the same line 😛

File: src/api/webhooks.js [hunk 4]
```diff
@@ -96,8 +87,6 @@ export class WebhooksApi extends BaseApi {
      */
     delete(webhookId) {
         const url = this.getFullURL('webhooks', webhookId);
-        return this.validateAuthHeaders().then((authHeaders) => {
-            return http('DELETE', url, undefined, authHeaders);
-        });
+        return this.request('DELETE', url);
```
Comment (mspensieri) [inline src/api/webhooks.js]: How come in some instances you pass empty object for the third param, but here it's ok to leave it out?

Comment (lemieux) [inline src/api/webhooks.js]: third params is the data, DELETE doesn't send data... but that probably should be handled in `http` instead

Comment (Mario54) [inline src/api/webhooks.js]: Hmm looks like it still works if I remove the empty objects in the third param

Comment (lemieux) [inline src/api/webhooks.js]: yeah I don't see why you should need it, I'd try smooch core with a script just to make sure

Comment (Mario54) [inline src/api/webhooks.js]: @mspensieri Fixed!

@lemieux Yup I tried it with a script and with the web widget and nothing breaks.

File: src/api/base.js (hunk 1 above)
Comment (dannytranlx) [inline src/api/base.js]: nice

**Comment by dannytranlx** [comment]
👍

**Comment by mspensieri** [comment]
:+1:

**Pull request merged**